### PR TITLE
Implement Package.get and Package.copy

### PR DESCRIPTION
**Pull request by dimaryaz**

### CODE REVIEW

File: api/python/t4/snapshots.py [hunk 1]
```diff
@@ -147,8 +182,8 @@ def download_bytes_from_snapshot(src, snapshothash):
     return download_bytes(src, version=obj_rec['VersionId'])
 
 class PhysicalKeyType(Enum):
```
Comment (akarve) [inline api/python/t4/snapshots.py]: Do we actually need `PhysicalKeyType` at all if we're using URIs?

Comment (akarve) [reply api/python/t4/snapshots.py]: I think we want an enum of URI schemes instead? maybe that is for a separate PR.

File: api/python/t4/snapshots.py [hunk 2]
```diff
@@ -333,18 +378,28 @@ def get(self, logical_key):
...
-            # TODO: dispatch on target to deserialize
-            raise NotImplementedError
 
-        raise NotImplementedError
+        return deserialize_obj(data, target), entry.meta.get('user_meta')
 
-    def get_files(self, logical_key, path):
+    def copy(self, logical_key, dest):
         """
         Gets objects from logical_key inside the package and saves them to path.
```
Comment (akarve) [inline api/python/t4/snapshots.py]: "saves them to dest"?

File: api/python/t4/snapshots.py [hunk 3]
```diff
@@ -333,18 +378,28 @@ def get(self, logical_key):
...
         # TODO: verify hash
-        if 'target' in entry.meta:
-            # TODO: dispatch on target to deserialize
-            raise NotImplementedError
 
-        raise NotImplementedError
+        return deserialize_obj(data, target), entry.meta.get('user_meta')
 
-    def get_files(self, logical_key, path):
+    def copy(self, logical_key, dest):
```
Comment (akarve) [inline api/python/t4/snapshots.py]: does `api.py` need to change with this? i.e. are we ready to make `get_file` and `put_file` into `copy`? OK to punt on this.

File: api/python/t4/snapshots.py (hunk 1 above)
Comment (dimaryaz) [reply api/python/t4/snapshots.py]: Probably not

File: api/python/t4/snapshots.py (hunk 3 above)
Comment (dimaryaz) [reply api/python/t4/snapshots.py]: Yeah, we probably want to do that - but wanted to make sure that URIs actually make sense

File: api/python/t4/snapshots.py [hunk 4]
```diff
@@ -45,6 +47,39 @@ def _parse_snapshot_path(snapshot_file_key):
...
+            # TODO: metadata
+            shutil.copyfile(src_url.path, dest_url.path)
+        elif dest_url.scheme == 's3':
+            upload_file(src_url.path, dest_url.netloc + dest_url.path, meta)
+        else:
+            raise NotImplementedError
+    elif src_url.scheme == 's3':
+        if dest_url.scheme == 'file':
+            # TODO: metadata
+            download_file(src_url.netloc + src_url.path, dest_url.path)
```
Comment (stevededalus) [inline api/python/t4/snapshots.py]: I think we need params here so we can download the versioned files?

File: api/python/t4/snapshots.py [hunk 5]
```diff
@@ -45,6 +47,39 @@ def _parse_snapshot_path(snapshot_file_key):
...
+        elif dest_url.scheme == 's3':
+            copy_object(src_url.netloc + src_url.path, dest_url.netloc + dest_url.path, meta)
```
Comment (stevededalus) [inline api/python/t4/snapshots.py]: maybe params here too?

File: api/python/t4/snapshots.py [hunk 6]
```diff
@@ -160,12 +195,22 @@ def hash_file(readable_file):
...
+        if url.scheme != 'file':
+            raise ValueError('Unexpected scheme: %r' % url.scheme)
+        with open(url.path, 'rb') as fd:
+            return fd.read()
+    elif ty == PhysicalKeyType.S3:
+        if url.scheme != 's3':
+            raise ValueError('Unexpected scheme: %r' % url.scheme)
+        return download_bytes(url.netloc + url.path)[0]
+    else:
+        assert False
```
Comment (stevededalus) [inline api/python/t4/snapshots.py]: Raise an exception with unsupport physical key type (maybe "unsupported storage" is more user friendly)?

File: api/python/t4/snapshots.py [hunk 7]
```diff
@@ -333,18 +378,28 @@ def get(self, logical_key):
...
+
+        data = read_physical_key(physical_key)
+
         # TODO: verify hash
-        if 'target' in entry.meta:
-            # TODO: dispatch on target to deserialize
-            raise NotImplementedError
 
-        raise NotImplementedError
+        return deserialize_obj(data, target), entry.meta.get('user_meta')
```
Comment (stevededalus) [inline api/python/t4/snapshots.py]: update the return type in the docstring

**Pull request merged**